File: backend/routes/demoRequests.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const DemoRequest = require('../models/DemoRequest');
const { verifyAdminToken } = require('./adminAuth');
const router = express.Router();

// Get all demo requests (admin)
router.get('/', verifyAdminToken, async (req, res) => {
  try {
    const { status, priority, search } = req.query;
    const query = {};
    if (status) query.status = status;
    if (priority) query.priority = priority;
    if (search) {
      query.$or = [
        { firstName: { $regex: search, $options: 'i' } },
        { lastName: { $regex: search, $options: 'i' } },
        { email: { $regex: search, $options: 'i' } },
        { company: { $regex: search, $options: 'i' } }
      ];
    }
    const requests = await DemoRequest.find(query).sort({ createdAt: -1 });
    const statusCounts = await DemoRequest.aggregate([{ $group: { _id: '$status', count: { $sum: 1 } } }]);
    res.json({ 
      requests, 
      statusCounts: statusCounts.reduce((acc, curr) => { acc[curr._id] = curr.count; return acc; }, {}) 
    });
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

// Get single demo request (admin)
router.get('/:id', verifyAdminToken, async (req, res) => {
  try {
    const request = await DemoRequest.findById(req.params.id);
    if (!request) return res.status(404).json({ message: 'Request not found' });
    res.json({ request });
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

// Update demo request (admin)
router.put('/:id', verifyAdminToken, async (req, res) => {
  try {
    const request = await DemoRequest.findByIdAndUpdate(req.params.id, req.body, { new: true });
    if (!request) return res.status(404).json({ message: 'Request not found' });
    res.json({ message: 'Request updated', request });
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

// Add note (admin)
router.post('/:id/notes', verifyAdminToken, [
  body('content').trim().notEmpty()
], async (req, res) => {
  try {
    const request = await DemoRequest.findByIdAndUpdate(
      req.params.id,
      { $push: { notes: { content: req.body.content, addedBy: req.admin.email, addedAt: new Date() } } },
      { new: true }
    );
    if (!request) return res.status(404).json({ message: 'Request not found' });
    res.json({ message: 'Note added', request });
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

// Delete demo request (admin)
router.delete('/:id', verifyAdminToken, async (req, res) => {
  try {
    await DemoRequest.findByIdAndDelete(req.params.id);
    res.json({ message: 'Request deleted' });
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

// PUBLIC: Submit demo request
router.post('/public/submit', [
  body('firstName').trim().notEmpty(),
  body('lastName').trim().notEmpty(),
  body('email').isEmail().normalizeEmail(),
  body('company').trim().notEmpty()
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ message: 'Validation failed', errors: errors.array() });
    }
    const demoRequest = new DemoRequest(req.body);
    await demoRequest.save();
    res.status(201).json({ message: 'Demo request submitted successfully' });
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

module.exports = router;

File: backend/routes/demo_requests.py
from fastapi import APIRouter, HTTPException, Request
from datetime import datetime
from bson import ObjectId

router = APIRouter()

async def get_demo_stats(db):
    total = await db.demo_requests.count_documents({})
    pending = await db.demo_requests.count_documents({"status": "new"})
    return {"total": total, "pending": pending}

@router.get("/")
async def get_demo_requests(request: Request, status: str = None, limit: int = 50, offset: int = 0):
    db = request.app.state.db
    
    query = {}
    if status:
        query["status"] = status
    
    cursor = db.demo_requests.find(query).sort("createdAt", -1).skip(offset).limit(limit)
    requests = await cursor.to_list(length=limit)
    total = await db.demo_requests.count_documents(query)
    
    for req in requests:
        req["id"] = str(req.pop("_id"))
    
    return {"requests": requests, "total": total}

@router.get("/{request_id}")
async def get_demo_request(request_id: str, request: Request):
    db = request.app.state.db
    
    demo_req = await db.demo_requests.find_one({"_id": ObjectId(request_id)})
    if not demo_req:
        raise HTTPException(status_code=404, detail="Demo request not found")
    
    demo_req["id"] = str(demo_req.pop("_id"))
    return demo_req

@router.post("/")
async def create_demo_request(request: Request):
    db = request.app.state.db
    data = await request.json()
    
    demo_doc = {
        "firstName": data.get("firstName"),
        "lastName": data.get("lastName"),
        "email": data.get("email"),
        "company": data.get("company"),
        "phone": data.get("phone"),
        "message": data.get("message"),
        "companySize": data.get("companySize"),
        "preferredDate": data.get("preferredDate"),
        "status": "new",
        "createdAt": datetime.utcnow()
    }
    
    result = await db.demo_requests.insert_one(demo_doc)
    
    return {
        "message": "Demo request submitted successfully",
        "status": "received",
        "id": str(result.inserted_id)
    }

@router.put("/{request_id}")
async def update_demo_request(request_id: str, request: Request):
    db = request.app.state.db
    data = await request.json()
    
    update_data = {k: v for k, v in data.items() if k not in ["_id", "id"]}
    update_data["updatedAt"] = datetime.utcnow()
    
    result = await db.demo_requests.update_one(
        {"_id": ObjectId(request_id)},
        {"$set": update_data}
    )
    
    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail="Demo request not found")
    
    return {"message": "Demo request updated"}

@router.delete("/{request_id}")
async def delete_demo_request(request_id: str, request: Request):
    db = request.app.state.db
    
    result = await db.demo_requests.delete_one({"_id": ObjectId(request_id)})
    
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="Demo request not found")
    
    return {"message": "Demo request deleted"}
